pkg/compute/bidder.go

```diff
@@ -5,69 +5,104 @@ import (
-		callback:      params.Callback,
+		nodeID:           params.NodeID,
+		store:            params.Store,
+		getApproveURL:    params.GetApproveURL,
+		callback:         params.Callback,
+		semanticStrategy: params.SemanticStrategy,
+		resourceStrategy: params.ResourceStrategy,
 	}
 }
 
-func (b Bidder) RunBidding(ctx context.Context, execution store.Execution) {
-	// ask the bidding strategy if we should bid on this job
-	bidStrategyRequest := bidstrategy.BidStrategyRequest{
-		NodeID:   b.nodeID,
-		Job:      execution.Job,
-		Callback: b.getApproveURL(),
-	}
+func (b Bidder) RunBidding(ctx context.Context, request AskForBidRequest, usageCalc capacity.UsageCalculator) {
+	go func() {
+		// ask the bidding strategy if we should bid on this job
+		bidStrategyRequest := bidstrategy.BidStrategyRequest{
+			NodeID:   b.nodeID,
+			Job:      request.Job,
+			Callback: b.getApproveURL(),
+		}
 
-	response, err := b.doBidding(ctx, bidStrategyRequest, execution.ResourceUsage)
-	if err != nil {
-		err = multierr.Append(err, b.store.UpdateExecutionState(ctx, store.UpdateExecutionStateRequest{
-			ExecutionID: execution.ID,
-			NewState:    store.ExecutionStateFailed,
-			Comment:     err.Error(),
-		}))
+		response, resourceUsage, err := b.doBidding(ctx, bidStrategyRequest, usageCalc)
+		if err != nil {
+			b.callback.OnComputeFailure(ctx, ComputeError{
+				RoutingMetadata: RoutingMetadata{
+					// TODO double check these.
+					SourcePeerID: b.nodeID,
+					TargetPeerID: request.SourcePeerID,
+				},
+				ExecutionMetadata: ExecutionMetadata{
+					ExecutionID: request.ExecutionID,
+					JobID:       request.JobID,
+				},
+				Err: err.Error(),
+			})
+
+			log.Ctx(ctx).Error().Err(err).Msg("Error running bid strategy")
+			return
+		}
 
-		b.callback.OnComputeFailure(ctx, ComputeError{
+		result := BidResult{
 			RoutingMetadata: RoutingMetadata{
 				SourcePeerID: b.nodeID,
-				TargetPeerID: execution.RequesterNodeID,
+				TargetPeerID: request.SourcePeerID,
 			},
-			ExecutionMetadata: NewExecutionMetadata(execution),
-			Err:               err.Error(),
-		})
+			ExecutionMetadata: ExecutionMetadata{
+				ExecutionID: request.ExecutionID,
+				JobID:       request.JobID,
+			},
+			Accepted: response.ShouldBid,
+			Reason:   response.Reason,
+		}
 
-		log.Ctx(ctx).Error().Err(err).Msg("Error running bid strategy")
-		return
-	}
+		// if we are not bidding and not wait return a response, we can't do this job.
+		if !response.ShouldBid && !response.ShouldWait {
+			b.callback.OnBidComplete(ctx, result)
+		}
 
-	b.ReturnBidResult(ctx, execution, response)
+		// if we are bidding or waiting create an execution
+		if response.ShouldWait || response.ShouldBid {
+			execution := store.NewExecution(request.ExecutionID, request.Job, request.SourcePeerID, *resourceUsage)
+			if err := b.store.CreateExecution(ctx, *execution); err != nil {
+				log.Ctx(ctx).Error().Err(err).Msg("Unable to create execution state")
+				return
+			}
```

Comment:
We previously were creating an Execution in all cases, and now we only create one if we intend to bid. I can see some advantages from an auditing and retries perspective in the compute node remembering what it has declined to bid on. Should we be keeping the previous behaviour, but just not populating the resource usage?

Reply:
True. I made this change as creating an execution for each bid, regardless of it being accepted or rejected, probably won't scale as the network grows. I also wanted to avoid adding "incomplete executions" (i.e. without resource usage date) in the execution store. But this decision is ultimately up to the compute node implementer. However, it seems unreasonable (imo) from a protocol level to expect a compute nodes to keep execution states for jobs they are not bidding on/rejecting. (e.g. A Requester shouldn't expect compute nodes to maintain state for bids they are rejecting.) 

With respect to retries, compute nodes could use their execution store to track jobs they have accepted or rejected, and again this is up to their implementation, but it would probably be more efficient for them to use a separate store/cache specifically for this purpose. (assuming we want to swiftly reject previously reject jobs?). I suppose there could be a case for keeping this state when a compute node initially rejects a job, but then later via a retry accepts it if more resources became available. 

With respect to auditing, I would expect an audit of bid accept/reject to happen at the Requester level given it's the "thingy" that ultimately cares about what has been accepted/rejected. If we want to make the case that compute nodes can be audited in this way we could use a filesystem journal managed by each compute nodes (again up to them).

If you feel strongly about keeping this state in the compute nodes I am happy to revert this change.